### PR TITLE
Add support for AssumeRoleWithWebIdentity

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add support for WebIdentity

## Motivation and Context
Implements and supports https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html 

## Regression
No its a new feature

## How Has This Been Tested?
This is a work in progress but testing details will be posted soon

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6985-027148e/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6985-027148e/gopath/src/github.com/minio/minio
++ cd 6985-027148e/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6985-027148e/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6985/head:pr6985
++ cd 6985-027148e/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6985
++ minikube ip
+ minikube_ip=192.168.39.96
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.96
2018/12/17 00:43:44 wrote public.crt
2018/12/17 00:43:44 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6985-027148e/gopath
+ make -C 6985-027148e/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/sts-handlers.go:51:56: sts.AssumeRoleWithWebIdentity undefined (type *stsAPIHandlers has no field or method AssumeRoleWithWebIdentity)
cmd/sts-handlers.go:140:6: (*stsAPIHandlers).AssumeRoleWithClientGrants redeclared in this block
	previous declaration at cmd/sts-handlers.go:64:6
make: *** [build] Error 2

```

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6985-a44590a/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6985-a44590a/gopath/src/github.com/minio/minio
++ cd 6985-a44590a/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6985-a44590a/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6985/head:pr6985
++ cd 6985-a44590a/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6985
++ minikube ip
+ minikube_ip=192.168.39.96
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.96
2018/12/18 15:21:37 wrote public.crt
2018/12/18 15:21:37 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6985-a44590a/gopath
+ make -C 6985-a44590a/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/sts-handlers.go:51:56: sts.AssumeRoleWithWebIdentity undefined (type *stsAPIHandlers has no field or method AssumeRoleWithWebIdentity)
cmd/sts-handlers.go:140:6: (*stsAPIHandlers).AssumeRoleWithClientGrants redeclared in this block
	previous declaration at cmd/sts-handlers.go:64:6
make: *** [build] Error 2

```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
This PR is ready for review @nitisht @kannappanr

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by vadmeste** [comment]
@harshavardhana, the temporary credentials issued after a successful oauth have equal power as the owner ?

**Comment by harshavardhana** [comment]
No never temp credentials have no power unless specified by policy either through OPA or a `policy` claim added as part of the JWT token received from your web identity provider. Requirements are same as the client grants credential.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
Here is how you can test with WSO2 , once you have installed and logged into the WSO2 management console you can navigate to 

### Add a local claim
https://localhost:9443/carbon/identity-claim-mgt/add-local-claim.jsp
policy

### Add an external claim to map OIDC
https://localhost:9443/carbon/identity-claim-mgt/add-external-claim.jsp
oidc-policy

### Now edit OIDC scope list for `openid` 
https://localhost:9443/carbon/oauth/add-oidc-claims.jsp?scopeName=openid

add-openid-scope

### Add claims to be mapped under service provider
https://localhost:9443/carbon/application/configure-service-provider.jsp?spName=minio-app

click

Once selected `policy` click update below.

### Now Create a new user
Add a new user and update the provider to match something like this
harsha

## Final
After that once you have a client id and client secrets, you can perform a password based OpenID Connect
```
curl -s -u NZLOOFRSluw9RfIkuHGqfk1HFp4a:0Z4VTG8uJBSekn42HE40DK9vQb4a -k -d "grant_type=password&username=harsha&password=harsha123" -H "Content-Type:application/x-www-form-urlencoded" https://localhost:9443/oauth2/token?scope=openid | jq .
{
  "access_token": "eyJ4NXQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJraWQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJoYXJzaGFAY2FyYm9uLnN1cGVyIiwiYXVkIjoiTlpMT09GUlNsdXc5UmZJa3VIR3FmazFIRnA0YSIsIm5iZiI6MTU0NjUwMjgzOSwiYXpwIjoiTlpMT09GUlNsdXc5UmZJa3VIR3FmazFIRnA0YSIsInNjb3BlIjoib3BlbmlkIiwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5NDQzXC9vYXV0aDJcL3Rva2VuIiwiZXhwIjoxNTQ2NTA2NDM5LCJpYXQiOjE1NDY1MDI4MzksImp0aSI6ImQ0YWRkOWQyLWEyNjItNGJjMi04MTNmLTZmOGIxYWYxMzY3NyIsInBvbGljeSI6IndyaXRlb25seSJ9.awrZyvcG832U90T87fPZLrRK97E2z13fiQ5iVvsGKom4Yrh2Z6J_ioOpPuXc6Rbz3gaIbHBICx65REM2AYKAa8GlXXDfeWGxmZEbqPglitcln_zdA3bsG0fGF4lpmshgVTbvD-JpmOfZVwyZosLi6i2YMiUvjQ7HYtQ4U0gvznrpnCbuKi0mp0VcfohsSKVQmclN3GlkNA80rt8r3wss7mSJb30iRwfFtARHUpwRDRnHMfQqOuBarxxHU2Od2xBcTGA6UVjaECJ-ypm-4IbU-K3w0T2DN7ITLNjHVjz_VC1qwiajO_TvMloexf5_83oLa--NnDLP1pycoSBirs7rDQ",
  "refresh_token": "7ba478c1-463e-3efe-90b2-1584ccd85250",
  "scope": "openid",
  "id_token": "eyJ4NXQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJraWQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJhbGciOiJSUzI1NiJ9.eyJhdF9oYXNoIjoiVENKeHBrYWt6V3R6Q09pSW04RV9hdyIsImF1ZCI6Ik5aTE9PRlJTbHV3OVJmSWt1SEdxZmsxSEZwNGEiLCJzdWIiOiJoYXJzaGEiLCJuYmYiOjE1NDY1MDI4MzksImF6cCI6Ik5aTE9PRlJTbHV3OVJmSWt1SEdxZmsxSEZwNGEiLCJhbXIiOlsicGFzc3dvcmQiXSwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5NDQzXC9vYXV0aDJcL3Rva2VuIiwiZXhwIjoxNTQ2NTA2NDM5LCJpYXQiOjE1NDY1MDI4MzksInBvbGljeSI6IndyaXRlb25seSJ9.OHFzN7MP9Yoa71N7LhyBGZfOXNWlhVbnHTxolRgXd32pqjsfKCw12Va7MOCL0ModRTMHXZRL9YYfqJ0ZDzdLaJm51ewAcPs1HsdwScb6tjkja83wEdnFxqVrQ3U5x_1DB6MRwNvy2lKoFj9AgNcLs2HYKffYinEHKNSU4VlqwjgGarRjXbjQRyNSRcMvYmcBub2w7AriT1I9j2vbHA1mf7WqNsRqlmPZAbcm9PQCQkrCd0clhg2bDjdDt1igFqtnpgbRyFBkKjixvsF_OeJ8dGVOYRaqDgBAprhaaOSsEk0IIt7vKJBXlXMswrf8QQJa8X_uJNkST6txzYgkCsoOcw",
  "token_type": "Bearer",
  "expires_in": 3600
}
```

The `id_token` what we should be using here to connect to Minio WebIdentity API and get credentials 

```
~ jwt-decode eyJ4NXQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJraWQiOiJOVEF4Wm1NeE5ETXlaRGczTVRVMVpHTTBNekV6T0RKaFpXSTRORE5sWkRVMU9HRmtOakZpTVEiLCJhbGciOiJSUzI1NiJ9.eyJhdF9oYXNoIjoiVENKeHBrYWt6V3R6Q09pSW04RV9hdyIsImF1ZCI6Ik5aTE9PRlJTbHV3OVJmSWt1SEdxZmsxSEZwNGEiLCJzdWIiOiJoYXJzaGEiLCJuYmYiOjE1NDY1MDI4MzksImF6cCI6Ik5aTE9PRlJTbHV3OVJmSWt1SEdxZmsxSEZwNGEiLCJhbXIiOlsicGFzc3dvcmQiXSwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5NDQzXC9vYXV0aDJcL3Rva2VuIiwiZXhwIjoxNTQ2NTA2NDM5LCJpYXQiOjE1NDY1MDI4MzksInBvbGljeSI6IndyaXRlb25seSJ9.OHFzN7MP9Yoa71N7LhyBGZfOXNWlhVbnHTxolRgXd32pqjsfKCw12Va7MOCL0ModRTMHXZRL9YYfqJ0ZDzdLaJm51ewAcPs1HsdwScb6tjkja83wEdnFxqVrQ3U5x_1DB6MRwNvy2lKoFj9AgNcLs2HYKffYinEHKNSU4VlqwjgGarRjXbjQRyNSRcMvYmcBub2w7AriT1I9j2vbHA1mf7WqNsRqlmPZAbcm9PQCQkrCd0clhg2bDjdDt1igFqtnpgbRyFBkKjixvsF_OeJ8dGVOYRaqDgBAprhaaOSsEk0IIt7vKJBXlXMswrf8QQJa8X_uJNkST6txzYgkCsoOcw
```

> {
>   "x5t": "NTAxZmMxNDMyZDg3MTU1ZGM0MzEzODJhZWI4NDNlZDU1OGFkNjFiMQ",
>   "kid": "NTAxZmMxNDMyZDg3MTU1ZGM0MzEzODJhZWI4NDNlZDU1OGFkNjFiMQ",
>   "alg": "RS256"
> }
> {
>   "at_hash": "TCJxpkakzWtzCOiIm8E_aw",
>   "aud": "NZLOOFRSluw9RfIkuHGqfk1HFp4a",
>   "sub": "harsha",
>   "nbf": 1546502839,
>   "azp": "NZLOOFRSluw9RfIkuHGqfk1HFp4a",
>   "amr": [
>     "password"
>   ],
>   "iss": "https://localhost:9443/oauth2/token",
>   "exp": 1546506439,
>   "iat": 1546502839,
>   **"policy": "writeonly"**
> }

**Comment by harshavardhana** [comment]
In-case you have trouble testing please ask me @vadmeste @Praveenrajmani

File: cmd/sts-handlers.go [hunk 1]
```diff
@@ -45,82 +44,104 @@ func registerSTSRouter(router *mux.Router) {
 	// AssumeRoleWithClientGrants
 	stsRouter.Methods("POST").HandlerFunc(httpTraceAll(sts.AssumeRoleWithClientGrants)).
 		Queries("Action", "AssumeRoleWithClientGrants").
+		Queries("Version", stsAPIVersion).
 		Queries("Token", "{Token:.*}")
 
-}
+	// AssumeRoleWithWebIdentity
+	stsRouter.Methods("POST").HandlerFunc(httpTraceAll(sts.AssumeRoleWithWebIdentity)).
```
Comment (vadmeste) [inline cmd/sts-handlers.go:51]: Probably, stsRouter should avoid logging here, since this can leak some information

**Review by vadmeste — COMMENTED** [review]
Tested & LGTM but one minor comment

File: cmd/sts-handlers.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/sts-handlers.go:51]: There is no information leak @vadmeste - one can never access or validate anything here. The request itself is anonymous so I don't see removing this saves us from anything.

**Review by vadmeste — DISMISSED** [review]
LGTM & tested with Google + OPA

**Comment by harshavardhana** [comment]
Fixed a typo @Praveenrajmani @vadmeste - PTAL

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#c1b124d4-3e74-414f-b949-ada25318df78) |
<a name="c1b124d4-3e74-414f-b949-ada25318df78"></a>
### 6985-d237b99/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30327
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bee31c99a9fd:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 57 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 55 seconds
(5/13) Running aws-sdk-ruby tests ... done in 33 seconds
(6/13) Running mc tests ... done in 1 minutes and 49 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 5 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) (Minio API responded with message=Multiple disks failures, unable to write data.) ---> Minio.Exceptions.MinioException: Minio API responded with message=Multiple disks failures, unable to write data.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 488
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 234
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

File: docs/sts/web-identity.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  "openid": {
+    "jwks": {
+      "url": "https://www.googleapis.com/oauth2/v3/certs"
+     }
+   }
+}
+```
+
+Testing with an example
+> Visit [Google Developer Console](https://console.cloud.google.com) under Project, APIs, Credentials to get your OAuth2 client credentials. Add `http://localhost:8080/google/callback` as a valid OAuth2 Redirect URL.
```
Comment (Praveenrajmani) [inline docs/sts/web-identity.md]: Typo - The url should be `http://localhost:8080/oauth2/callback`

Comment (harshavardhana) [reply docs/sts/web-identity.md]: Will fix it..

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by Praveenrajmani — APPROVED** [review]
Tested with Google + OPA. LGTM

**Pull request merged**